server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -116,4 +205,158 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
             }
         }
     }
+
+    private static class ExtractFromSource extends IndexRouting {
+        private final String indexName;
+        private final FilterPath[] include;
+
+        ExtractFromSource(int routingNumShards, int routingFactor, String indexName, List<String> routingPaths) {
+            super(routingNumShards, routingFactor);
+            this.indexName = indexName;
+            this.include = FilterPath.compile(Set.copyOf(routingPaths));
+        }
+
+        @Override
+        public int indexShard(
+            boolean idIsAutoGenerated,
+            String id,
+            @Nullable String routing,
+            XContentType sourceType,
+            BytesReference source
+        ) {
+            if (idIsAutoGenerated == false) {
+                throw new IllegalArgumentException(error("indexing with a specified id"));
+            }
+            if (routing != null) {
+                throw new IllegalArgumentException(error("indexing with a specified routing"));
+            }
+            try {
+                try (
+                    XContentParser parser = sourceType.xContent()
+                        .createParser(
+                            NamedXContentRegistry.EMPTY,
+                            DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                            source.streamInput(),
+                            include,
+                            null
+                        )
+                ) {
+                    parser.nextToken(); // Move to first token
+                    if (parser.currentToken() == null) {
+                        throw new IllegalArgumentException("Error extracting routing: source didn't contain any routing fields");
+                    }
+                    int hash = extractObject(parser);
+                    ensureExpectedToken(null, parser.nextToken(), parser);
+                    return hashToShardId(hash);
+                }
+            } catch (IOException | ParsingException e) {
+                throw new IllegalArgumentException("Error extracting routing: " + e.getMessage(), e);
+            }
+        }
+
+        private static int extractObject(XContentParser source) throws IOException {
+            ensureExpectedToken(Token.FIELD_NAME, source.nextToken(), source);
+            String firstFieldName = source.currentName();
+            source.nextToken();
+            int firstHash = extractItem(source);
+            if (source.currentToken() == Token.END_OBJECT) {
+                // Just one routing key in this object
+                // Use ^ like Map.Entery's hashcode
+                return Murmur3HashFunction.hash(firstFieldName) ^ firstHash;
+            }
+            String[] fields = new String[] { firstFieldName, null };
+            int[] hashes = new int[] { firstHash, 0 };
```

Comment:
I wonder if this is better than just having a `List<FieldAndHash> fields = new ArrayList<>()`? AFAIK, we will often have more than 2 fields. And allocating an object per (field,hash) pair seems unproblematic, we are parsing json (and indexing) after all here.

Reply:
Here's the `routing_path` that I put together for some k8s data:
```
    "routing_path": [
      "kubernetes.container.id",
      "kubernetes.event.involved_object.uid",
      "kubernetes.node.name",
      "kubernetes.pod.name",
      "kubernetes.system.container",
      "kubernetes.volume.name",
      "metricset.name"
    ]
```

In my data there are always `metricset.name` and at least one of the others.

I think it'll be quite common for a "leaf" to have a single value. So I like the single-value optimization. But I think you are right about the `List<FieldAndHash>` for the more than one value case.